### PR TITLE
Replace update all confirm with dialog

### PR DESCRIPTION
**Pull request by Misiu**
Replace browser confirm dialog with proper dialog.
Replace this:
image

with this:
image

### CODE REVIEW

File: src/update-all/update-all-dialog.ts [hunk 1]
```diff
@@ -1,24 +1,42 @@
...
+  @property() public configuration!: string;
+
   protected render() {
     return html`
-      <esphome-process-dialog
-        heading="Update All"
-        .type=${"update-all"}
-        @closed=${this._handleClose}
-      ></esphome-process-dialog>
+      <mwc-dialog heading="Update All" @closed=${this._handleClose} open>
```
Comment (balloob) [inline src/update-all/update-all-dialog.ts:14]: Please create a generic confirm dialog that works with promises.

```ts
if (await confirm("Update All", "Do you want to update all devices", "Update All")) {
    ...
}
```

And then call that from the openUpdateAllDialog function

Comment (Misiu) [reply src/update-all/update-all-dialog.ts:14]: I think that this is a bit beyond my knowledge, but I'll search in HA Frontend repo and maybe will be able to propose something.

**Comment by Misiu** [comment]
OK, I have some progress:
image
This can be now reused (also has a destructive style).
I've also added red color for the Delete device in the card menu:
image

File: src/components/esphome-header-menu.ts [hunk 2]
```diff
@@ -74,21 +75,28 @@ export class ESPHomeHeaderMenu extends LitElement {
...
   private _handleEditSecrets() {
     openEditDialog(SECRETS_FILE);
   }
 
-  private _handleOverflowAction(ev: CustomEvent<ActionDetail>) {
+  private async _handleOverflowAction(ev: CustomEvent<ActionDetail>) {
     switch (ev.detail.index) {
       case 0:
-        this._handleUpdateAll();
+        await this._handleUpdateAll();
```
Comment (Misiu) [inline src/components/esphome-header-menu.ts:99]: VS Code suggested this, not sure if _handleEditSecrets should be async too

**Comment by Misiu** [comment]
I can do other places (delete device) in the next PR.
I used HA Frontend code as a reference, but I didn't include all possible dialogs (alert, prompt) because I wanted to do a small PR.

File: src/components/esphome-header-menu.ts (hunk 2 above)
Comment (balloob) [reply src/components/esphome-header-menu.ts:99]: awaiting it only helps with finding the stack trace when things blow up, but it doesn't really matter here. We don't care about the result.

File: src/devices/configured-device-card.ts [hunk 3]
```diff
@@ -136,7 +136,7 @@ class ESPHomeConfiguredDeviceCard extends LitElement {
             <mwc-list-item>Download YAML</mwc-list-item>
             <mwc-list-item>Rename</mwc-list-item>
             <mwc-list-item>Clean Build Files</mwc-list-item>
-            <mwc-list-item>Delete</mwc-list-item>
+            <mwc-list-item class="destructive">Delete</mwc-list-item>
```
Comment (balloob) [inline src/devices/configured-device-card.ts:139]: For a future PR, if you're looking for things to contribute, let's add icons 👍

File: src/dialogs/index.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      //   } else if (ev.detail.dialog === "cancel") {
+      //     origCancel && origCancel();
+      //   }
+      //   resolve();
+      // });
+      // document.body.appendChild(dialog);
+      // dialog.showDialog({
+      //   ...dialogParams,
+      //   ...extra,
+      // });
```
Comment (balloob) [inline src/dialogs/index.ts:54]:
```suggestion
```

File: src/update-all/index.ts [hunk 5]
```diff
@@ -1,6 +1,8 @@
-const preload = () => import("./update-all-dialog");
+const preload = () => import("./update-all-process-dialog");
```
Comment (balloob) [inline src/update-all/index.ts:1]: let's revert this rename as the folder has not changed either.

File: src/dialogs/confirmation-dialog.ts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+      <mwc-dialog
+        heading=${this._params?.title || ""}
+        @closed=${this._handleClose}
+        open
+      >
+        ${this._params?.text && html`<div>${this._params?.text}</div>`}
+        <mwc-button
+          slot="secondaryAction"
+          no-attention
+          label=${this._params?.dismissText || "Cancel"}
```
Comment (balloob) [inline src/dialogs/confirmation-dialog.ts:30]:
```suggestion
          label=${this._params.dismissText || "Cancel"}
```

File: src/dialogs/confirmation-dialog.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    this._params = params;
+  }
+
+  protected render(): TemplateResult {
+    if (!this._params) {
+      return html``;
+    }
+    return html`
+      <mwc-dialog
+        heading=${this._params?.title || ""}
```
Comment (balloob) [inline src/dialogs/confirmation-dialog.ts:22]:
```suggestion
        heading=${this._params.title || ""}
```

File: src/dialogs/confirmation-dialog.ts [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        @closed=${this._handleClose}
+        open
+      >
+        ${this._params?.text && html`<div>${this._params?.text}</div>`}
```
Comment (balloob) [inline src/dialogs/confirmation-dialog.ts:26]:
```suggestion
        ${this._params.text && html`<div>${this._params.text}</div>`}
```

File: src/dialogs/confirmation-dialog.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        <mwc-button
+          slot="secondaryAction"
+          no-attention
+          label=${this._params?.dismissText || "Cancel"}
+          @click=${this._dismiss}
+        ></mwc-button>
+        <mwc-button
+          slot="primaryAction"
+          label=${this._params?.confirmText || "Yes"}
```
Comment (balloob) [inline src/dialogs/confirmation-dialog.ts:35]:
```suggestion
          .label=${this._params.confirmText || "Yes"}
```

File: src/dialogs/index.ts [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  title?: string;
+  confirmText?: string;
+  dismissText?: string;
+  confirm?: () => void;
+  cancel?: () => void;
+  destructive?: boolean;
+}
+
+export const showConfirmationDialog = (
+  dialogParams: ConfirmationDialogParams
```
Comment (balloob) [inline src/dialogs/index.ts:14]: Don't accept a `confirm` and `cancel` here. Only return value should be used.

File: src/dialogs/index.ts [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+) =>
```
Comment (balloob) [inline src/dialogs/index.ts:15]:
```suggestion
): Promise<boolean> =>
```

File: src/dialogs/index.ts [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+  new Promise((resolve) => {
+    const origCancel = dialogParams.cancel;
+    const origConfirm = dialogParams.confirm;
```
Comment (balloob) [inline src/dialogs/index.ts:18]: Let's remove this. We don't use it and we shouldn't

File: src/dialogs/index.ts (hunk 4 above)
Comment (Misiu) [reply src/dialogs/index.ts]: done

File: src/dialogs/confirmation-dialog.ts (hunk 6 above)
Comment (Misiu) [reply src/dialogs/confirmation-dialog.ts]: done

File: src/dialogs/confirmation-dialog.ts (hunk 7 above)
Comment (Misiu) [reply src/dialogs/confirmation-dialog.ts]: done

File: src/dialogs/confirmation-dialog.ts (hunk 8 above)
Comment (Misiu) [reply src/dialogs/confirmation-dialog.ts]: done

File: src/dialogs/confirmation-dialog.ts (hunk 9 above)
Comment (Misiu) [reply src/dialogs/confirmation-dialog.ts]: done

File: src/dialogs/index.ts (hunk 11 above)
Comment (Misiu) [reply src/dialogs/index.ts]: done

File: src/update-all/index.ts (hunk 5 above)
Comment (Misiu) [reply src/update-all/index.ts]: done

File: src/dialogs/index.ts (hunk 10 above)
Comment (Misiu) [reply src/dialogs/index.ts:12]: done

File: src/dialogs/index.ts (hunk 12 above)
Comment (Misiu) [reply src/dialogs/index.ts]: done

File: src/devices/configured-device-card.ts (hunk 3 above)
Comment (Misiu) [reply src/devices/configured-device-card.ts:139]: done

File: src/components/esphome-header-menu.ts (hunk 2 above)
Comment (Misiu) [reply src/components/esphome-header-menu.ts]: I've removed the await keyword.

**Comment by Misiu** [comment]
ready for the next round of reviews :)

File: src/dialogs/confirmation-dialog.ts [hunk 13]
```diff
@@ -8,9 +8,14 @@ import { classMap } from "lit/directives/class-map.js";
 @customElement("esphome-confirmation-dialog")
 class ESPHomeConfirmationDialog extends LitElement {
   @state() private _params?: ConfirmationDialogParams;
+  @state() private _resolve?: (value: boolean | PromiseLike<boolean>) => void;
```
Comment (balloob) [inline src/dialogs/confirmation-dialog.ts:11]: You can change these to be. They are always set if dialog is open. Only keep `if (!this._params) {` inside `render`
```suggestion
  @state() private _params!: ConfirmationDialogParams;
  @state() private _resolve!: (value: boolean | PromiseLike<boolean>) => void;
```

File: src/dialogs/confirmation-dialog.ts [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+    if (!this._params) {
+      return html``;
+    }
+    return html`
+      <mwc-dialog
+        .heading=${this._params.title || ""}
+        @closed=${this._handleClose}
+        open
+      >
+        ${this._params.text && html`<div>${this._params.text}</div>`}
```
Comment (balloob) [inline src/dialogs/confirmation-dialog.ts:31]: This is not React where boolean values are filtered out. Use ternary statements.
```suggestion
        ${this._params.text ? html`<div>${this._params.text}</div>` : ""}
```

File: src/dialogs/confirmation-dialog.ts [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+        <mwc-button
+          slot="secondaryAction"
+          no-attention
+          .label=${this._params.dismissText || "Cancel"}
+          @click=${this._dismiss}
```
Comment (balloob) [inline src/dialogs/confirmation-dialog.ts:36]:
```suggestion
          @click=${this._dismiss}
          dialogAction="dismiss"
```

File: src/update-all/index.ts [hunk 16]
```diff
@@ -1,6 +1,6 @@
 const preload = () => import("./update-all-dialog");
 
-export const openUpdateAllDialog = () => {
+export const openUpdateAllProcessDialog = () => {
```
Comment (balloob) [inline src/update-all/index.ts:3]: can you rename this dialog back

File: src/update-all/update-all-dialog.ts [hunk 17]
```diff
@@ -23,6 +23,6 @@ class ESPHomeUpdateAllDialog extends LitElement {
 
 declare global {
   interface HTMLElementTagNameMap {
-    "esphome-update-all-dialog": ESPHomeUpdateAllDialog;
+    "esphome-update-all-process-dialog": ESPHomeUpdateAllDialog;
```
Comment (balloob) [inline src/update-all/update-all-dialog.ts:26]: revert this. it doesn't match the custom element decorator.

File: src/update-all/update-all-dialog.ts (hunk 1 above)
Comment (Misiu) [reply src/update-all/update-all-dialog.ts]: done

File: src/dialogs/confirmation-dialog.ts (hunk 13 above)
Comment (Misiu) [reply src/dialogs/confirmation-dialog.ts]: done

File: src/dialogs/confirmation-dialog.ts (hunk 14 above)
Comment (Misiu) [reply src/dialogs/confirmation-dialog.ts:31]: oops

File: src/dialogs/confirmation-dialog.ts (hunk 15 above)
Comment (Misiu) [reply src/dialogs/confirmation-dialog.ts]: done

File: src/dialogs/confirmation-dialog.ts [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+        ></mwc-button>
+        <mwc-button
+          slot="primaryAction"
+          .label=${this._params.confirmText || "Yes"}
+          class=${classMap({
+            destructive: this._params.destructive || false,
+          })}
+          dialogAction="close"
```
[the remark this answers is not in the document]
Comment (Misiu) [reply src/dialogs/confirmation-dialog.ts]: done

File: src/dialogs/confirmation-dialog.ts [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+          @click=${this._confirm}
+        ></mwc-button>
+      </mwc-dialog>
+    `;
+  }
+
+  private _handleClose() {
```
[the remark this answers is not in the document]
Comment (Misiu) [reply src/dialogs/confirmation-dialog.ts]: done

File: src/update-all/update-all-dialog.ts (hunk 17 above)
Comment (Misiu) [reply src/update-all/update-all-dialog.ts]: done

File: src/update-all/index.ts (hunk 16 above)
Comment (Misiu) [reply src/update-all/index.ts]: done

File: src/update-all/index.ts [hunk 20]
```diff
@@ -1,6 +1,7 @@
 const preload = () => import("./update-all-dialog");
 
 export const openUpdateAllDialog = () => {
-  preload();
-  document.body.append(document.createElement("esphome-update-all-dialog"));
+  preload().then(() => {
```
Comment (Misiu) [inline src/update-all/index.ts:4]: preload is a Promise, so I think we should use then or await it. Please confirm, so I'll do the same in other places

Comment (balloob) [reply src/update-all/index.ts:4]: it's a promise but you don't need to await it. Elements are automatically upgraded when loaded.

Comment (Misiu) [reply src/update-all/index.ts:4]: So I should remove `then` or can we leave it as is?

Comment (balloob) [reply src/update-all/index.ts:4]: revert the change.

Comment (Misiu) [reply src/update-all/index.ts]: done

File: src/dialogs/confirmation-dialog.ts [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+            destructive: this._params.destructive || false,
+          })}
+          dialogAction="confirm"
+        ></mwc-button>
+      </mwc-dialog>
+    `;
+  }
+
+  private _handleClose(e: CustomEvent): void {
+    if (!this._resolve) return;
```
Comment (balloob) [inline src/dialogs/confirmation-dialog.ts:51]: Not needed
```suggestion
```

**Review by balloob — APPROVED** [review]

**Pull request merged**